qubes-gui-client/src/buffer.rs

```diff
@@ -198,6 +198,11 @@ impl<T: VchanMock + 'static> RawMessageStream<T> {
     ///
     /// Fails if there is an I/O error on the vchan.
     pub fn write(&mut self, buf: &[u8]) -> Result<(), vchan::Error> {
+        #[cfg(not(test))]
+        match self.state {
+            ReadState::Error | ReadState::Connecting | ReadState::Negotiating => return Ok(()),
+            _ => {}
+        }
```

Comment:
Changing the behavior of the code depending on whether we're running tests or not is very sus and likely to blow up in our face in the future. I'd at least expect a very good justification in the comment.

The logic itself is kinda confusing too. I'd expect that `ReadState::Error` returns `Err` and otherwise the message gets buffered to be sent once the connection gets established.

Reply:
In that case, perhaps `panic` to not mask the bug?

Reply:
It’s not a bug, but rather intentional behavior.  That said, I suspect this code is far, _far_ too complex.

Reply:
I mean, if somebody *doesn't* poll `needs_reconnect`, this will break silently, won't it?

Reply:
Yes, it will.  A callback-based API is much less error-prone here.